# DTSC 580: Data Manipulation
## Assignment: LEGO Project
### Name: 

![LegoSchema](Lego_Schema.PNG)

## Overview

In this assignment, you will load the various tables from the [Rebrickable](https://rebrickable.com/) website and perform analysis on the data.  Rebrickable is a fantastic website that is used to show you which LEGO sets you can build from the sets and parts you already own.  They will even show you which parts you are missing to complete the new set and suggest sets that you could purchase to get those parts.  

## Data

The data has been downloaded directly from their website and includes files for the tables shown in the schema above.  Your first task will be to go through the data and review the schema to understand how the tables all fit together.  There is no data dictionary that I am aware of, but you should be able to understand this data by reviewing the files and schema and visiting their [Download](https://rebrickable.com/downloads/) page and [API](https://rebrickable.com/api/) documentation.  **Note that you must use the files downloaded from Brightspace for this assignment and not download new files from the Rebrickable website.**

## Assignment

In this assignment, you will create a number of functions to complete certain tasks.  This is not needed if you were working on this project on your own, but since we want to see if you can select the data via Pandas/Numpy, we will use those functions in the automatic grading in CodeGrade.  Without this, students could try to look up the answers in the `csv` files instead of wrangling the data with Python/Pandas, which would defeat the purpose of this class.

In addition, we would strongly suggest that you merge data sets as appropriate for the given task.  For example, you might merge the `sets` and `themes` DataFrames to answer one question, and then merge `sets`, `inventories`, `inventory_minifigs`, and `minifigs` to answer another.  This will make your code much easier to write and will be key in working with data across multiple tables.  In fact, I was able to create each function with only one line of code because of the merging of certain data sets.  Also note that you may not use all the tables in this assignment.  This is also an important skill for a Data Scientist to know when to use data and what data is unnecessary.

When complete, save your notebook as `lego.ipynb` and submit to CodeGrade for automatic grading.

## Note

<u>Show Work</u>

Remember that you must show your work.  Students submissions are spot checked manually throughout the term to verify that they are not hard coding the answer from looking only in the file or in CodeGrade's expected output.  If this is seen, the student's answer will be manually marked wrong and their grade will be changed to reflect this. 

For example, if the answer to Q1, the mean of a specific column, is 22:
```
# correct way
Q1 = df['column_name'].mean()

# incorrect way
Q1 = 22 
```

Let's start out by importing some standard imports.

In [65]:
# common imports
import pandas as pd
import numpy as np

# Do not change this option; This allows the CodeGrade auto grading to function correctly
pd.set_option('display.max_columns', None)

## Import Data

Load the various Lego datasets below calling them the same as the name of the table as seen in the schema.  Note that you may or may not have to use all the data sets.  One of the things that all Data Scientists and Analysts have to decide when working on a project is what data is useful for a task and what data is not.  Remember to look at the schema above to see how all the data sets are connected.

The autograder in CodeGrade will have the `csv` files in the same location where the notebook is run.  Because of this, please make sure your local files are in the same location as your notebook.

In [66]:
# import Lego data sets here
# make sure that files are in the same folder as your notebook
inventories = pd.read_csv('inventories.csv')
inventory_sets = pd.read_csv('inventory_sets.csv')
sets = pd.read_csv('sets.csv')
themes = pd.read_csv('themes.csv')
inventory_minifigs = pd.read_csv('inventory_minifigs.csv')
minifigs = pd.read_csv('minifigs.csv')
inventory_parts = pd.read_csv('inventory_parts.csv')
part_categories = pd.read_csv('part_categories.csv')
parts = pd.read_csv('parts.csv')
colors = pd.read_csv('colors.csv')
part_relationships = pd.read_csv('part_relationships.csv')
elements = pd.read_csv('elements.csv')

## Exploration

Take time to look at all of the data sets using `.head()`, `.info()`, `.describe()`, etc to familiarize yourself with the data sets.  As you look at them, notice how the various tables connect together.

In [67]:
### ENTER CODE HERE ###
inventory_sets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3413 entries, 0 to 3412
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   inventory_id  3413 non-null   int64 
 1   set_num       3413 non-null   object
 2   quantity      3413 non-null   int64 
dtypes: int64(2), object(1)
memory usage: 80.1+ KB


## Assignment

Answer all the questions below using the available data.  There are multiple ways to get some of the answers.  I would suggest that you attempt to merge some data sets as needed to assist with getting answers.  This will make your tasks a lot easier and your code will be cleaner.

Also, you will be creating functions to get the answers to some of these questions.  This will make it much easier for you to get answers to similar questions in the future and CodeGrade will be checking your functions using different inputs.  

**Q1:** Create a function called `data_count` that takes as input the data set (for example: `sets`) and returns the total number of rows in the data (as an integer).  Nothing else should be returned.

In [68]:
### ENTER CODE HERE ###

def data_count(df):
    return len(df)


**Code Check:**  If you call your `data_count` function using the `sets` DataFrame, you should see that there are 18,576 sets represented in the data.

In [69]:
### CODE CHECK ###

data_count(sets)

18576

**Q2:** What is the average number of parts for a Lego set (rounded to the nearest integer)?  Save this as `Q2`.

In [70]:
### ENTER CODE HERE ###

flt = sets['num_parts'].mean()

Q2 = round(flt)

Q2

161

**Q3:** What is the median number of parts per set (output as an integer)?  Save this as `Q3`.

In [71]:
### ENTER CODE HERE ###
flt2 = sets['num_parts'].median()

Q3= round(flt2)

Q3


37

You should see that the median number of parts is much lower than the average number of parts per set.  This tells us that there are some Lego sets with a very high number of parts that brings the average up.

**Q4:** What is the largest number of parts that a set has? Save this as `Q4`.

In [72]:
### ENTER CODE HERE ###

Q4 = sets['num_parts'].max()

Q4

11695

**Q5:** Given the set with the largest number of parts calculated above, select its row from the `sets` DataFrame.  Save this as `Q5`.  Do not reset the index.

In [73]:
### ENTER CODE HERE ###

Q5 = sets.loc[sets['num_parts'] == 11695]

Q5

set_num       name  year  theme_id  num_parts
3867  31203-1  World Map  2021       709      11695

**Q6:** Create a function called `select_set_row` where you pass it the `set_num` (as a string) and the function returns the respective row of the DataFrame from `sets`. Do not reset the index.

In [74]:
### ENTER CODE HERE ###

sets2 = sets.rename(columns={'name_sets': 'name'})

def select_set_row(set_num):
    return sets2.loc[sets['set_num'] == set_num]


**Q7:** Create a function called `select_set_numparts` that takes as input a specific number of parts (as an integer) and selects all the rows in the `sets` DataFrame that contain those number of parts.  Return the DataFrame sorted by `year` and then by `set_num`.  Do not reset the index.

In [75]:
### ENTER CODE HERE ###
def select_set_numparts(num_parts):
    return sets2.loc[sets2['num_parts'] == num_parts].sort_values(by=['year', 'set_num'])



**Code Check:** Call the `select_set_numparts()` function to select all sets with `0` number of parts.  There should be 3,146 rows returned.

In [76]:
### CODE CHECK ###

select_set_numparts(0)

set_num                                               name  \
11061          700.1-2                     Extra-Large Gift Set (Mursten)   
17563         b55dk-01               Lego Mursten - System i Leg Byggebog   
17564         b56de-01                                    System im Spiel   
11072          700.2-2                                       Gift Package   
11078         700.3A-2                                       Gift Package   
...                ...                                                ...   
15619         854195-1                                    Ernie Key Chain   
15620         854197-1                               Demogorgon Key Chain   
17170  9780744054583-1  DC Comics Super Heroes: Character Encyclopedia...   
17193  9780794447557-1                               City: Color The City   
17200  9780794448172-1           DC Comics Super Heroes: Ready For Action   

       year  theme_id  num_parts  
11061  1953       366          0  
17563  1955       497          0  
17564  1956       497          0  
11072  1957       366          0  
11078  1957       372          0  
...     ...       ...        ...  
15619  2022       503          0  
15620  2022       503          0  
17170  2022       497          0  
17193  2022       497          0  
17200  2022       497          0  

[3146 rows x 5 columns]

**Q8:** 
- The oldest set listed was made in what year?  Save the year as `Q8A` (as an integer).  
- The newest set listed was made in what year? Save the year as `Q8B` (as an integer).

In [77]:
### ENTER CODE HERE ###
Q8A = sets['year'].min()

Q8A



1949

In [78]:
### ENTER CODE HERE ###
Q8B = sets['year'].max()

Q8B

2022

**Q9:** 
- Create a function called `select_set_year` that takes as input a year (as an integer) and returns the rows from the `sets` DataFrame that matches that year.  Sort the DataFrame by `set_num`.  Do not reset the index.
- Use the above function to return a DataFrame for rows that match the minimum year (calculated as `Q8A` above) and call this `Q9A`.  
- Then use the above function to return a DataFrame for rows that match the maximum year (calculated as `Q8B` above) and call this `Q9B`.

In [79]:
### ENTER CODE HERE ###
def select_set_year(year):
    return sets2.loc[sets2['year'] == year].sort_values(by=['set_num'])

In [80]:
### ENTER CODE HERE ###

Q9A = select_set_year(1949)

Q9A

Q9B = select_set_year(2022)

Q9B

set_num                                               name  \
336            10297-1                                     Boutique Hotel   
600            10781-1                          Spider-Man’s Techno Trike   
601            10782-1                      Hulk vs. Rhino Truck Showdown   
602            10783-1                        Spider-Man at Doc Ock’s Lab   
603            10784-1                     Spider-Man Webquarters Hangout   
...                ...                                                ...   
17170  9780744054583-1  DC Comics Super Heroes: Character Encyclopedia...   
17193  9780794447557-1                               City: Color The City   
17200  9780794448172-1           DC Comics Super Heroes: Ready For Action   
17207  9780794449247-1  DC Comics Super Heroes: Gotham City’s New Defe...   
17726    DOTS-SERIES-6                    DOTS Collectible Tiles Series 6   

       year  theme_id  num_parts  
336    2022       155       3068  
600    2022       706         59  
601    2022       706        110  
602    2022       706        131  
603    2022       706        155  
...     ...       ...        ...  
17170  2022       497          0  
17193  2022       497          0  
17200  2022       497          0  
17207  2022       497          7  
17726  2022       688         16  

[183 rows x 5 columns]

**Q10:** What is the median year represented in the `sets` data?  Save this as `Q10` (as an integer).

In [81]:
### ENTER CODE HERE ###

year_median = sets['year'].median()

Q10 = round(year_median)

Q10

2010

Notice that the first set was back in 1949 but over half of the sets were after the median year. This could be an indication that the number of sets has increased greatly over the past number of years.

**Q11:** What is the average number of sets listed per year?  Save this as `Q11` (as an integer).

In [82]:
### ENTER CODE HERE ###

avg_sets_year = sets.groupby('year')['set_num'].count().mean()

Q11 = round(avg_sets_year)

Q11

258

**Q12:** 
- Create a function called `theme_by_year` that takes as input a year (as an integer) and shows the theme ids and theme names (listed in order by theme id) that were in sets that year.  
- The column names must be `id` and `name_themes` (to differentiate between the name of a theme and the name of a set) in that order.  
- The index should be reset and go from `0` to `n-1`.  
- Each theme should only be listed once even if it appeared in more than one set from that year -- duplicate themes should be based on theme id and not name since there are some themes with the same name but with a different id.
- Hint:  It will help if you were to think about merging appropriate DataFrames to help you get this answer.

In [83]:
### ENTER CODE HERE ###
themes.rename(columns={"name": "name_themes"}, inplace=True)

sets_themes = pd.merge(sets, themes, how="inner", left_on="theme_id", right_on="id")

def theme_by_year(year):
    p1 = sets_themes.loc[sets_themes['year'] == year].sort_values(by='theme_id').drop_duplicates(subset='theme_id')
    return p1[['id', 'name_themes']].reset_index(drop=True)

theme_by_year(1960)

id   name_themes
0  371  Supplemental
1  497         Books
2  513       Classic

**Code Check:** Call the `theme_by_year()` function on the year `1960`.  Your output should look as follows:

| | id | name_themes|
|-|---|---|
|0|371|Supplemental|
|1|497|Books|
|2|513|Classic|

**Optional Practice:** Call the `theme_by_year()` function using your birth year to see the Lego themes that were popular when you were born.

In [84]:
### CODE CHECK / OPTIONAL PRACTICE ###
sets_themes

set_num                                name  year  theme_id  \
0            001-1                               Gears  1965         1   
1            002-1      4.5V Samsonite Gears Motor Set  1965         1   
2           1030-1      TECHNIC I: Simple Machines Set  1985         1   
3           1038-1                  ERBIE the Robo-Car  1985         1   
4           1039-1                Manual Control Set 1  1986         1   
...            ...                                 ...   ...       ...   
18571  M20-2566-10  Modulex Box - 10x Window 1 x 6 x 6  1963       716   
18572  M20-2575-10  Modulex Box - 10x Window 1 x 7 x 5  1963       716   
18573  M20-2576-10  Modulex Box - 10x Window 1 x 7 x 6  1963       716   
18574  M20-2586-10  Modulex Box - 10x Window 1 x 8 x 6  1963       716   
18575  WEETABIX1-1                     Weetabix Castle  1970       414   

       num_parts   id name_themes  parent_id  
0             43    1     Technic        NaN  
1              3    1     Technic        NaN  
2            211    1     Technic        NaN  
3            120    1     Technic        NaN  
4             39    1     Technic        NaN  
...          ...  ...         ...        ...  
18571         10  716     Modulex        NaN  
18572         10  716     Modulex        NaN  
18573         10  716     Modulex        NaN  
18574         10  716     Modulex        NaN  
18575        471  414      Castle      411.0  

[18576 rows x 8 columns]

**Q13:** What theme has the most number of sets associated with it?  Save the theme name as `Q13` (as a string).

In [85]:
### ENTER CODE HERE ###

theme_most_sets = sets_themes.groupby('name_themes')['set_num'].count().idxmax()

Q13 = theme_most_sets

Q13

'Gear'

**Q14:** 
- Create a function called `theme_by_name` that takes as input a theme name and outputs all the sets (in `set_num` order) that had that theme.  
- The DataFrame should include the following columns in this order and with these exact names: `set_num`,`name_sets`,`year`,`num_parts`, `theme_id`, and `name_themes`.  
- The index should go from `0` to `n-1`.

In [86]:
### ENTER CODE HERE ###
theme_name = sets_themes.rename(columns={'name': 'name_sets'})

theme_name2 =theme_name.drop(['id', 'parent_id'], axis=1)

theme_name3 = theme_name2.reindex(['set_num','name_sets','year', 'num_parts', 'theme_id', 'name_themes'], axis=1)

def theme_by_name(name_themes):
    return theme_name3.loc[theme_name3['name_themes'] == name_themes].sort_values(by='set_num').reset_index(drop=True)

**Code Check:** Call the `theme_by_name()` function using the theme name from `Q13`.  You should return 1,904 sets that match that theme name.

In [87]:
### CODE CHECK ###

theme_by_name('Gear')

set_num                                          name_sets  \
0              11138-1                            Bionicle Barraki Wallet   
1       119645111436-1                Classic Space Astronaut Plush (Red)   
2       119645111504-1               Classic Space Astronaut Plush (Blue)   
3       119645111740-1              Classic Space Astronaut Plush (White)   
4       119645111818-1               Classic Space Astronaut Plush (Pink)   
...                ...                                                ...   
1899           WATCH-3                   Watch System Voyager (UFO Space)   
1900         WIIGAME-1                              Bionicle Heroes - Wii   
1901         WIIGAME-2  Indiana Jones: The Original Adventures Video G...   
1902  WOODENELEPHANT-1  Original Wooden Elephant Pull-Along Toy and Print   
1903          ZX8000-1                               ZX 8000 LEGO Sneaker   

      year  num_parts  theme_id name_themes  
0     2008          0       501        Gear  
1     2021          0       501        Gear  
2     2021          0       501        Gear  
3     2021          0       501        Gear  
4     2021          0       501        Gear  
...    ...        ...       ...         ...  
1899  1997         29       501        Gear  
1900  2007          0       501        Gear  
1901  2008          0       501        Gear  
1902  2021          0       501        Gear  
1903  2020          0       501        Gear  

[1904 rows x 6 columns]

**Q15:** Create a function called `theme_by_setnum` that takes as input the set_num (as a string) and returns the respective theme name as a string.

In [88]:
### ENTER CODE HERE ###

def theme_by_setnum(set_num):
    p2 = theme_name3.loc[theme_name3['set_num'] == set_num]
    return p2['name_themes'].max()

**Code Check:** Visit the [Sets page](https://rebrickable.com/sets/) from the Rebrickable website and find a set that you like.  Call your `theme_by_setnum()` function on that set and make sure the correct theme name is output.

In [ ]:
### CODE CHECK ###

**Q16:** Find all the sets named 'Millennium Falcon'. Of these sets, find the set with the largest number of parts.  You **do not** need to save this as `Q16`.  Instead you will use this information for later questions.

In [ ]:
### ENTER CODE HERE ###

**Q17:** 
- How many minifigs are contained in the set from `Q16`?  Save this as `Q17` (as an integer).  
- Note that there are sometimes multiple versions of a set as shown in the `inventories` table.  This will not matter for this question as there is only one version for the respective set but this should be noted for future questions.
- Remember to show your work or your answer will manually get marked as incorrect.

In [89]:
### ENTER CODE HERE ###
figs_sets = pd.merge(sets, inventories , how="inner", left_on="set_num", right_on="set_num")
figs_sets2 = pd.merge(figs_sets, inventory_minifigs , how="inner", left_on="id", right_on="inventory_id")
figs_sets3 = pd.merge(figs_sets2, minifigs , how="inner", left_on="fig_num", right_on="fig_num")

Q17 = figs_sets3[figs_sets3['set_num'] == '75192-1']['fig_num'].count()
Q17

8

**Q18:** 
- Create a function called `get_minifigs` that takes as input a set number (as a string) and a version number (as an integer) and outputs all minifigs from that set / version combination.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `set_num`, `fig_num`, `fig_name`, `fig_num_parts`.  
- The index should go from 0 to n-1.

In [90]:
### ENTER CODE HERE ###
minifigs.rename(columns={'name': 'fig_name', 'num_parts':'fig_num_parts'}, inplace=True)
figs_vents = pd.merge(sets, inventories , how="inner", left_on="set_num", right_on="set_num")
figs_vents2 = pd.merge(figs_vents, inventory_minifigs , how="inner", left_on="id", right_on="inventory_id")
figs_vents3 = pd.merge(figs_vents2, minifigs , how="inner", left_on="fig_num", right_on="fig_num")

def get_minifigs(set_num, version):
    p3 = figs_vents3.loc[figs_vents3['set_num'] == set_num].sort_values(by='fig_num')
    p4 = p3.loc[figs_vents3['version'] == version].sort_values(by='fig_num')
    return p4[['set_num', 'fig_num', 'fig_name', 'fig_num_parts']].reset_index(drop=True)

**Q19:** Using the above function, output information about the minifigs from the set discussed in `Q16`.  Save this output as `Q19`.

In [91]:
### ENTER CODE HERE ###
Q19 = get_minifigs('75192-1', 1)

Q19

set_num     fig_num                                           fig_name  \
0  75192-1  fig-001714                              Chewbacca, Dark Brown   
1  75192-1  fig-001810                             Finn, Black Undershirt   
2  75192-1  fig-002057                                 Rey, Dark Tan Robe   
3  75192-1  fig-002514    C-3PO, Pearl Gold, Colorful Wires, Printed Legs   
4  75192-1  fig-002544                                               BB-8   
5  75192-1  fig-002549  Princess Leia, Hoth Outfit, White Jacket, Prin...   
6  75192-1  fig-002550  Han Solo, Dark Blue Jacket, Dark Brown Legs, P...   
7  75192-1  fig-002551                               Han Solo, Old, Angry   

   fig_num_parts  
0              3  
1              4  
2              4  
3              3  
4              2  
5              4  
6              4  
7              4

**Code Check:** Make sure that your `get_minifigs()` function works with both version 1 and version 2 of set 10217-1, Harry Potter's Diagon Alley.  Both versions should have 11 minifigs in the DataFrame.

In [ ]:
### CODE CHECK ###

**Q20:** 
- Create a function called `minifigs_from_themes` that takes as input a theme id (as an integer) and outputs all the minifigs contained in sets with that theme.  
- The DataFrame should be in `fig_num` order with the following columns/column names: `fig_num`, `fig_name`. 
- The index should go from 0 to n-1.  
- Exact duplicates should be removed so that only unique figs are included.

In [92]:
### ENTER CODE HERE ###
def minifigs_from_themes(theme_id):
    p5 = figs_vents3.loc[figs_vents3['theme_id'] == theme_id].sort_values(by='fig_num').drop_duplicates(subset='fig_num')
    return p5[['fig_num', 'fig_name']].reset_index(drop=True)
    

**Code Check:** Call the `minifigs_from_themes()` function using the theme number of the set discussed in `Q16`.  There should be 149 minifigs output for that respective theme.

In [93]:
### CODE CHECK ###
minifigs_from_themes(171)

fig_num                                           fig_name
0    fig-000003                              Assassin Droid, White
1    fig-000124                           Chewbacca, Reddish Brown
2    fig-000144  Luke Skywalker, White Robe, White Legs, Cheek ...
3    fig-000303           Clone Trooper, Phase I Armor, Brown Eyes
4    fig-000516  Darth Vader, Light Bluish Gray Skin, Scarred R...
..          ...                                                ...
144  fig-011942  AT-AT Driver - Light Bluish Grey Helmet, Sand ...
145  fig-011944  Snowtrooper, White Hips, Reddish Brown Head, F...
146  fig-011945  Snowtrooper, White Hips, Light Flesh Head, Female
147  fig-011946  Snowtrooper, White Hips, Light Flesh Head, Off...
148  fig-011957    Snowtrooper, White Hips, Light Flesh Head, Grin

[149 rows x 2 columns]

**Q21:** 
- Create a function called `sets_from_minifig` that takes as input a minifig number (as a string) and returns a list of sets a minifig has appeared in.  
- Return a DataFrame (sorted by `set_num` and then by `version`) with the following columns: `set_num`, `version`,`name_sets`, `fig_num`, `fig_name`.  
- The index should go from 0 to n-1.

In [94]:
### ENTER CODE HERE ###     
figs_vents4 = figs_vents3.rename(columns={"name": "name_sets"})
figs_vents4
def sets_from_minifig(fig_num):
    p7 = figs_vents4.loc[figs_vents4['fig_num'] == fig_num].sort_values(by = ['set_num', 'version'])
    return p7[['set_num', 'version', 'name_sets', 'fig_num', 'fig_name']].reset_index(drop=True)

**Code Check:** Call your `sets_from_minifig()` function on one of the Lego Friends minifigs, Andrea, who has figure number: `fig-001039`.  Your function should return a DataFrame that contains the following sets:

- Andrea on the Beach
- Birthday Party
- City Park Cafe

In [ ]:
### CODE CHECK ###

**Q22:** 

*Note:  Before you begin working on this question, please note that CodeGrade has a 1GB RAM limit.  If you see an `Exit -9` code when submitting your assignment, it is probably due to this RAM limit and previous students had an issue with this question causing too much memory usage.  The biggest thing to think about is to only use the tables that you need to answer for this question.  For instance, you do not need any of the minifigs data since this is not needed for the output and the merging of this data could cause the number of rows to "explode" in terms of the total count.  I understand that this can be annoying that we have to work around CodeGrade's limitations, but it is good practice in memory management.*

- Create a function called `set_parts` that takes as input the set number (as a string) and version number (as an integer) and outputs a DataFrame with each part in the set.  
- The DataFrame should have the following columns/column names: `set_num`, `version`, `name_sets`, `year`, `name_themes`, `part_num`, `name_part`, `name_color`,`quantity`,`is_spare`.  
- The Data should be sorted by `part_num` and then `name_color`.  
- The index should go from 0 to n-1.

In [95]:
### ENTER CODE HERE ###
sets3 = sets_themes.rename(columns={'name': 'name_sets'})
parts3 = parts.rename(columns={'name': 'name_part'})
colors3 = colors.rename(columns={'name': 'name_color'})

sets_parts = pd.merge(sets3, inventories , how="inner", left_on= "set_num", right_on="set_num")
sets_parts2 = pd.merge(sets_parts, inventory_parts , how="inner", left_on= "id_y", right_on="inventory_id")
sets_parts3 = pd.merge(sets_parts2, parts3 , how="inner", left_on= "part_num", right_on="part_num")
sets_parts4 = pd.merge(sets_parts3, colors3 , how="inner", left_on= "color_id", right_on="id")   
     

def set_parts(set_num,version):
    p8 = sets_parts4.loc[sets_parts4['set_num'] == set_num]
    p9 = p8.loc[sets_parts4['version'] == version].sort_values(by=['part_num', 'name_color'])
    return p9[['set_num', 'version', 'name_sets', 'year', 'name_themes', 'part_num', 'name_part', 'name_color', 'quantity', 'is_spare']].reset_index(drop=True)
                                                                                       



**Code Check**: Using the `set_parts()` function, output a list of parts for the set discussed in `Q16`. There should be a total of 730 rows in the DataFrame, which includes 676 parts and 54 spare parts. You can check the respective page on the rebrickable website to verify these results.

In [96]:
### CODE CHECK ###
set_parts('75192-1',1)

set_num  version          name_sets  year                name_themes  \
0    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
1    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
2    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
3    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
4    75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
..       ...      ...                ...   ...                        ...   
727  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
728  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
729  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
730  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   
731  75192-1        1  Millennium Falcon  2017  Ultimate Collector Series   

    part_num                                          name_part  \
0      10197  Technic Pin Connector Hub with 2 Perpendicular...   
1      10201            Bracket 1 x 2 - 1 x 4 [Rounded Corners]   
2      10247  Plate Special 2 x 2 with 1 Pin Hole [Complete ...   
3      10247  Plate Special 2 x 2 with 1 Pin Hole [Complete ...   
4      10314  Brick Curved 1 x 4 x 1 1/3 No Studs, Curved To...   
..       ...                                                ...   
727    99563  Tile Special 1 x 2 with Sloped Walls AKA Money...   
728    99780                     Bracket 1 x 2 - 1 x 2 Inverted   
729    99780                     Bracket 1 x 2 - 1 x 2 Inverted   
730    99781                              Bracket 1 x 2 - 1 x 2   
731    99784  Bar 1 x 12 with 1 x 2 Plate End with Hollow St...   

            name_color  quantity is_spare  
0     Dark Bluish Gray         3        f  
1    Light Bluish Gray        12        f  
2     Dark Bluish Gray         5        f  
3                White        22        f  
4                  Tan         3        f  
..                 ...       ...      ...  
727  Light Bluish Gray        70        f  
728              Black         6        f  
729   Dark Bluish Gray        25        f  
730  Light Bluish Gray         2        f  
731  Light Bluish Gray         3        f  

[732 rows x 10 columns]

  **Q23:**  Get a list of all unique color names, sort by color name, and output as a list.  Save this as `Q23`. 

In [98]:
### ENTER CODE HERE ###
Q23 = colors['name'].sort_values().unique().tolist()

Q23

['Aqua',
 'Black',
 'Blue',
 'Blue-Violet',
 'Bright Green',
 'Bright Light Blue',
 'Bright Light Orange',
 'Bright Light Yellow',
 'Bright Pink',
 'Brown',
 'Chrome Antique Brass',
 'Chrome Black',
 'Chrome Blue',
 'Chrome Gold',
 'Chrome Green',
 'Chrome Pink',
 'Chrome Silver',
 'Copper',
 'Coral',
 'Dark Azure',
 'Dark Blue',
 'Dark Blue-Violet',
 'Dark Bluish Gray',
 'Dark Brown',
 'Dark Flesh',
 'Dark Gray',
 'Dark Green',
 'Dark Orange',
 'Dark Pink',
 'Dark Purple',
 'Dark Red',
 'Dark Tan',
 'Dark Turquoise',
 'Earth Orange',
 'Fabuland Brown',
 'Fabuland Orange',
 'Flat Dark Gold',
 'Flat Silver',
 'Flesh',
 'Glitter Trans-Clear',
 'Glitter Trans-Dark Pink',
 'Glitter Trans-Light Blue',
 'Glitter Trans-Neon Green',
 'Glitter Trans-Orange',
 'Glitter Trans-Purple',
 'Glow In Dark Opaque',
 'Glow In Dark Trans',
 'Glow in Dark White',
 'Green',
 'Lavender',
 'Light Aqua',
 'Light Blue',
 'Light Bluish Gray',
 'Light Flesh',
 'Light Gray',
 'Light Green',
 'Light Lime',
 'Light 

**Q24:** How many total transparent colors are in the `colors` DataFrame?  Save this as `Q24` (as an integer).

In [99]:
### ENTER CODE HERE ###
Q24 = colors.loc[colors['is_trans'] == 't'].count().max()

Q24

38

**Q25:**  
- Create a function called `get_part_colors` that takes as input a part number (as a string) and returns a DataFrame of all colors a part has appeared in.  
- The DataFrame should be sorted by color id  
- The index should go from 0 to n-1.

In [102]:
### ENTER CODE HERE ###

colors2 = colors.rename(columns={'name_color': 'name', 'color_id':'id'})
parts_colors = pd.merge(inventory_parts, colors2 , how="inner", left_on= "color_id", right_on="id")

def get_part_colors(part_num):
    p10 = parts_colors.loc[parts_colors['part_num'] == part_num].sort_values(by='id')
    return p10[['id', 'name', 'rgb', 'is_trans' ]].drop_duplicates(subset = 'id').reset_index(drop=True)

id               name     rgb is_trans
0   0              Black  05131D        f
1   4                Red  C91A09        f
2  14             Yellow  F2CD37        f
3  15              White  FFFFFF        f
4  19                Tan  E4CD9E        f
5  71  Light Bluish Gray  A0A5A9        f
6  72   Dark Bluish Gray  6C6E68        f

## Optional Exercises
These exercises are optional and are not graded by CodeGrade.  They can be used as extra optional practice.

- What minifig has the most number of parts?
- What minifig shows up in the most number of sets?
- When was the first minifig included in a set?
- What set has the most number of minifigs
- What part is used the most for all sets?
- What color is used the most for all parts?
- What part / color combination is used the most for all sets?
- What part / color combination is used the least for all sets?
- Explore some of your favorite themes and which colors are used the most for those themes.
- Plot a line graph of the average number of parts per set per year.
